presto-main/src/main/java/com/facebook/presto/operator/scalar/MapConstructor.java

```diff
@@ -121,55 +120,52 @@ public BuiltInScalarFunctionImplementation specialize(BoundVariables boundVariab
-                METHOD_HANDLE.bindTo(mapType).bindTo(keyBlockEquals).bindTo(keyBlockHashCode).bindTo(keyIndeterminate),
-                Optional.of(instanceFactory));
+                METHOD_HANDLE.bindTo(mapType).bindTo(keyType).bindTo(valueType).bindTo(keyBlockEquals).bindTo(keyBlockHashCode).bindTo(keyIndeterminate),
+                Optional.empty());
     }
 
     @UsedByGeneratedCode
     public static Block createMap(
             MapType mapType,
+            Type keyType,
+            Type valueType,
             MethodHandle keyBlockEqual,
             MethodHandle keyBlockHashCode,
             MethodHandle keyIndeterminate,
-            State state,
             SqlFunctionProperties properties,
             Block keyBlock,
             Block valueBlock)
     {
         checkCondition(keyBlock.getPositionCount() == valueBlock.getPositionCount(), INVALID_FUNCTION_ARGUMENT, "Key and value arrays must be the same length");
-        PageBuilder pageBuilder = state.getPageBuilder();
-        if (pageBuilder.isFull()) {
-            pageBuilder.reset();
-        }
-
-        MapBlockBuilder mapBlockBuilder = (MapBlockBuilder) pageBuilder.getBlockBuilder(0);
+        MapBlockBuilder mapBlockBuilder = (MapBlockBuilder) mapType.createBlockBuilder(null, keyBlock.getPositionCount() * 2);
         BlockBuilder blockBuilder = mapBlockBuilder.beginBlockEntry();
+
         for (int i = 0; i < keyBlock.getPositionCount(); i++) {
             if (keyBlock.isNull(i)) {
-                // close block builder before throwing as we may be in a TRY() call
-                // so that subsequent calls do not find it in an inconsistent state
-                mapBlockBuilder.closeEntry();
                 throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "map key cannot be null");
             }
-            Object keyObject = readNativeValue(mapType.getKeyType(), keyBlock, i);
-            try {
-                if ((boolean) keyIndeterminate.invoke(keyObject, false)) {
-                    throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "map key cannot be indeterminate: " + mapType.getKeyType().getObjectValue(properties, keyBlock, i));
+
+            if (keyType.getJavaType() == Block.class) {
+                // If it's nto primitive or string, we need to look for nulls in the block.
+                Object keyObject = readNativeValue(mapType.getKeyType(), keyBlock, i);
+                try {
+                    if ((boolean) keyIndeterminate.invoke(keyObject, false)) {
+                        throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "map key cannot be indeterminate: " + mapType.getKeyType().getObjectValue(properties, keyBlock, i));
+                    }
+                }
+                catch (Throwable t) {
+                    throw internalError(t);
                 }
             }
-            catch (Throwable t) {
-                mapBlockBuilder.closeEntry();
```

Comment:
And this?

Reply:
Same this is a local object.- don't do unnecessary work